python/smqtk/utils/factors.py

```diff
@@ -6,62 +6,76 @@
-            / add dynamic programming concepts.
+    TODO: Add starting offset so as to be able to get primes within a range.
+
+    :param N: Value to get prime values less-than-or-equal this value.
+    :return: List of prime integer values <= N.
+
     """
     # TODO: Optimization - all primes after 2 are odd, so special case when
     #           N == 2, else loop over and deal only with odd numbers.
+    global _soe_c
 
     # in case we get N as a string or something
     N = float(N)
 
-    # maps non-prime number to an increment value defining the next value after
-    # the key that is not a prime.
-    not_primes = {}
-    primes = []
-
-    c = 2
-    while c <= N:
-        if c in not_primes:
-            i = not_primes[c]
-            # next non-prime number. Need to skip entries already in not_primes
-            # so we don't interfere with recorded key-interval pairings
-            nnp = c + i
-            while (nnp in not_primes) and (nnp < N):
-                nnp += i
-            # npp may become >N, meaning results of this interval have already
-            # been fully covered, and we can safely assign it out of the range
-            # of N as we know we won't be looking past N at the most in the map.
-            not_primes[nnp] = i
-            # Won't be looking at c again, so we can remove it from the map to
-            # save memory
-            del not_primes[c]
+    if _soe_c > N:
+        # We have already computed all primes up to N, so return cache-filled
+        # list.
+        primes = []
+        for p in _soe_prime_cache:
+            if p > N:
+                break
+            primes.append(p)
+        return primes
+
+    # We need to find more prime values, so start with the least known
+    # not-prime and sieve our way up.
+    while _soe_c <= N:
+        if _soe_c in _soe_not_prime_map:
+            # Get increment for the current non-prime value and record the next
+            # non-prime value, keeping this increment. Incrementing past N in
+            # order to main state in reused cache structure.
+            incr = _soe_not_prime_map[_soe_c]
+            nnp = _soe_c + incr
+            while (nnp in _soe_not_prime_map):
+                nnp += incr
+            _soe_not_prime_map[nnp] = incr
+            # Remove _soe_c from map as it has been considered now.
+            del _soe_not_prime_map[_soe_c]
         else:
-            primes.append(c)
-            # The lowest value that will
-            c_squared = c*c
-            if c_squared <= N:
-                not_primes[c_squared] = c
-        c += 1
+            # _soe_c is a prime value: record it and add square of _soe_c to
+            # not-prime-map (square of the prime is the first value not
+            # attaiable by multiplying preceeding prime value).
+            _soe_prime_cache.append(_soe_c)
+            _soe_not_prime_map[_soe_c*_soe_c] = _soe_c
+        _soe_c += 1
```

Comment:
I mean, this really shouldn't even be looking at even numbers at all. Just making the increment `+= 2` will help.